BuildScript/BuildTasks/TestConfigurationsFilter.cs

```diff
@@ -0,0 +1,201 @@
+      _enabledExecutionRuntimes = enabledExecutionRuntimes;
+      _supportedTargetRuntimes = supportedTargetRuntimes;
+      _enabledTargetRuntimes = enabledTargetRuntimes;
+      _supportedConfigurationIDs = supportedConfigurationIDs;
+      _enabledConfigurationIDs = enabledConfigurationIDs;
+    }
+
+    public IReadOnlyCollection<TestConfiguration> GetFilteredTestConfigurations (IReadOnlyCollection<TestConfiguration> testConfigurationsToFilter)
+    {
+      var errorsForUnsupportedMetadata = GetErrorsForUnsupportedMetadata (testConfigurationsToFilter).ToArray();
+      if (errorsForUnsupportedMetadata.Any())
+        throw new InvalidOperationException (string.Join (Environment.NewLine, errorsForUnsupportedMetadata));
+
+      return testConfigurationsToFilter
+          .Where (HasEnabledBrowser)
+          .Where (HasEnabledPlatform)
+          .Where (HasEnabledDatabaseSystem)
+          .Where (HasEnabledExecutionRuntime)
+          .Where (HasEnabledTargetRuntime)
+          .Where (HasEnabledConfigurationID)
+          .ToArray();
+    }
+
+    private IEnumerable<string> GetErrorsForUnsupportedMetadata (IEnumerable<TestConfiguration> testConfigurations)
+    {
+      foreach (var testConfiguration in testConfigurations)
+      {
+        if (!IsSupportedBrowser (testConfiguration.Browser))
+          yield return GetErrorMessageForUnsupportedFilterValue (TestConfigurationMetadata.Browser, testConfiguration.Browser);
+
+        if (!IsSupportedPlatform (testConfiguration.Platform))
+          yield return GetErrorMessageForUnsupportedFilterValue (TestConfigurationMetadata.Platform, testConfiguration.Platform);
+
+        if (!IsSupportedDatabaseSystem (testConfiguration.DatabaseSystem))
+          yield return GetErrorMessageForUnsupportedFilterValue (TestConfigurationMetadata.DatabaseSystem, testConfiguration.DatabaseSystem);
+
+        if (!IsSupportedExecutionRuntime (testConfiguration.ExecutionRuntime))
+          yield return GetErrorMessageForUnsupportedFilterValue (TestConfigurationMetadata.ExecutionRuntime, testConfiguration.ExecutionRuntime);
+
+        if (!IsSupportedTargetRuntime (testConfiguration.TargetRuntime))
+          yield return GetErrorMessageForUnsupportedFilterValue (TestConfigurationMetadata.TargetRuntime, testConfiguration.TargetRuntime);
+
+        if (!IsSupportedConfigurationID (testConfiguration.ConfigurationID))
+          yield return GetErrorMessageForUnsupportedFilterValue (TestConfigurationMetadata.ConfigurationID, testConfiguration.TargetRuntime);
+      }
+    }
+
+    private string GetErrorMessageForUnsupportedFilterValue (string metadata, string value)
+    {
+      return $"Cannot filter with unsupported {metadata} '{value}'";
+    }
+
+    private string GetErrorMessageForMissingEnabledValues (string metadata)
+    {
+      return $"No enabled {metadata}s were provided.";
+    }
+
+    private bool IsSupportedPlatform (string platform)
```

Comment:
should we keep those factored out or them inline? It comes down to readability and maintainability. given the amount to duplication and naming abstractions involved, it might be more efficient to do them inline. it might also end up as not-readable.

Reply:
We already shortly discussed this. While I would also inline them if they only abstracted the call to `Contains`, (there is no huge leap between SupportedX.Contains (Y) and IsSupportedX (Y)), `IsSupportedTargetRuntime` also does a format conversion, and to keep the calls symmetric I would leave them as they are right now